src/github.com/matrix-org/dendrite/roomserver/input/latest_events.go

```diff
@@ -66,69 +66,76 @@ func updateLatestEvents(
 		}
 	}()
 
-	err = doUpdateLatestEvents(db, updater, ow, roomNID, stateAtEvent, event, sendAsServer)
-	return
+	u := latestEventsUpdater{
+		db: db, updater: updater, ow: ow, roomNID: roomNID,
+		stateAtEvent: stateAtEvent, event: event, sendAsServer: sendAsServer,
+	}
+	return u.doUpdateLatestEvents()
 }
 
-func doUpdateLatestEvents(
-	db RoomEventDatabase,
-	updater types.RoomRecentEventsUpdater,
-	ow OutputRoomEventWriter,
-	roomNID types.RoomNID,
-	stateAtEvent types.StateAtEvent,
-	event gomatrixserverlib.Event,
-	sendAsServer string,
-) error {
+type latestEventsUpdater struct {
+	db                      RoomEventDatabase
+	updater                 types.RoomRecentEventsUpdater
+	ow                      OutputRoomEventWriter
+	roomNID                 types.RoomNID
+	stateAtEvent            types.StateAtEvent
+	event                   gomatrixserverlib.Event
+	sendAsServer            string
+	lastEventIDSent         string
+	latest                  []types.StateAtEventAndReference
+	removed                 []types.StateEntry
+	added                   []types.StateEntry
```

Comment:
You need to comment on the semantics/relationship between `removed` and `added`. They can be of unequal length (which I have derived from a random function), and we're going to compare the state-key-tuples of each element left-to-right and keep the 'lesser'. I don't know why we do this.